### ISSUE TITLE
Incorrect nullability warnings appear after pattern matching on certain 

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#50162]
**Version Used**: 

**Steps to Reproduce**:

Compile the following code
```cs
#nullable enable
class E
{
  public E F = new E();
}

class Test
{    
  void M1(E e)
  {
    switch (e)
    {
      case { F: { F: null } }: return;
    }
    e.F.ToString();
    e.F.F.ToString();
    e.F.F.F.ToString();
  }
    
  void M2(E e)
  {
    switch (e)
    {
      case { F: null }: return;
    }
    e.F.ToString();
    e.F.F.ToString();
    e.F.F.F.ToString();
    e.F.F.F.F.ToString();
  }
  
  void M3(E e)
  {
    switch (e)
    {
      case { F: { F: null } }: return;
    }
    e.F.ToString();
    e.F.F.ToString();
    e.F.F.F.ToString();
    e.F.F.F.F.ToString();
  }
    
  void M4(E e)
  {
    switch (e)
    {
      case { F: { F: { F: null } } }: return;
    }
    e.F.ToString();
    e.F.F.ToString();
    e.F.F.F.ToString(); // false warning
    e.F.F.F.F.ToString();
  }
}
```

**Expected Behavior**:
No warnings in all examples.
They are all the same but with different field depth. In all cases if the pure null check pattern matches the field the method exits and cannot dereference the field. Expressions after the switch statement are only reachable when the field is not nullable.

**Actual Behavior**:
`CS8602: Dereference of a possibly null reference.` is reported in `M4` but not any other methods.

**Notes**
I couldn't test whether the behavior is present with more deeply nested patterns since Roslyn crashes at the next level of nesting: https://github.com/dotnet/roslyn/issues.
However since the crash is likely caused by accessing a slot not allocated due to depth limit the behavior cannot be tested.

**Comment by bernd5** [comment]
It is a bug, but best would be if we could get a "stackoverflow-warning".

**Comment by RikkiGibson** [comment]
It looks like there was an issue intended to be linked in the **Notes** section.

**Comment by TessenR** [comment]
@RikkiGibson Sorry, I didn't notice the incorrect link in the buffer. The related issue is https://github.com/dotnet/roslyn/issues/50161. I've updated the **Notes** section as well.
